File: fun4.go
package main

import (
	"fmt"
)

func stratMaxScore(s, m int, posl []int) (int, int) {
	bestM := 0
	ourResBest := -999999

	for j := 1; j <= m; j++ {
		ourRes := 0
		for i := 0; i < j && s+i < len(posl); i++ {
			ourRes += posl[s+i]
		}
		if ourRes > ourResBest {
			ourResBest = ourRes
			bestM = j
		}
	}

	if bestM == 0 {
		bestM = 1
	}
	return bestM, ourResBest
}

func stratBestScore(s, m int, posl []int) (int, int) {
	bestM := 0
	ourResBest := -999999
	difference := -999999

	for j := 1; j <= m; j++ {
		ourRes := 0
		for i := 0; i < j && s+i < len(posl); i++ {
			ourRes += posl[s+i]
		}
		enemyM, enemyScore := stratMaxScore(s+j, m, posl)
		_ = enemyM // unused
		if ourRes-enemyScore > difference {
			difference = ourRes - enemyScore
			ourResBest = ourRes
			bestM = j
		}
	}

	if bestM == 0 {
		bestM = 1
	}
	return bestM, ourResBest
}

func fun4(m int) {
	var posl []int
	var i int

	fmt.Println("Введите последовательность чисел через пробел, для завершенния ввода введите любую букву и enter")
	for {
		_, err := fmt.Scan(&i)
		if err != nil {
			break
		}
		posl = append(posl, i)
	}

	pavelScore := 0
	vikaScore := 0
	s := 0

	for s < len(posl) {
		bestM, score := stratBestScore(s, m, posl)
		pavelScore += score
		s += bestM

		if s >= len(posl) {
			break
		}

		bestM, score = stratBestScore(s, m, posl)
		vikaScore += score
		s += bestM
	}

	if pavelScore > vikaScore {
		fmt.Println("1")
	} else {
		fmt.Println("0")
	}
}

File: fun4.cpp
#include "fun4.h"

#include <iostream>
#include <vector>
using namespace std;
pair<int, int> strat_max_score(int s, int m, vector<int>& posl){
    int best_m = 0;
    int our_res = 0;
    int our_res_best = -999999;
    for (int j = 1; j <= m; j++){
        our_res = 0;
        for (int i = 0; i < j && s+i < size(posl); i++){
            our_res += posl[s + i];
        }
        if (our_res > our_res_best){
            our_res_best = our_res;
            best_m = j;
        }
        
    }
    if (best_m == 0) best_m++;
    pair<int, int> result = {best_m, our_res_best};
    return result;
}

pair<int, int> strat_best_score(int s, int m, vector<int>& posl){ //находит наилучшую разницу между получением очков себе и не получением их противником
    int best_m = 0;
    int our_res = 0;
    int our_res_best = -999999;
    int difference = -999999;
    pair<int, int> enemy;
    for (int j = 1; j <= m; j++){
        our_res = 0;
        for (int i = 0; i < j && s+i < size(posl); i++){
            our_res += posl[s + i];
        }
        enemy = strat_max_score(s+j, m , posl);
        if (our_res - enemy.second > difference){
            difference = our_res - enemy.second;
            our_res_best = our_res;
            best_m = j;
        }
        
    }
    if (best_m == 0) best_m++;
    pair<int, int> result = {best_m, our_res_best};
    return result;
}


void fun4(int m){ // m - максимальная длинна хода
    vector<int> posl;
    int s = 0; // сколько чисел зачеркнули
    int i = 0;
    cout << "Введите последовательность чисел через пробел, для завершенния ввода введите любую букву и enter" << endl;
    while(cin >> i){
        posl.push_back(i);
    }
    int pavel_score = 0;
    int vika_score = 0;
    int turn = 0; // 0 - Pavel, 1 - Vika текущий ход
    pair<int, int> score;
    while (s < size(posl)){
        //cout << "Turn = " << turn << " ------------------" << endl;
        if (turn % 2 == 0){
            pair<int, int> score = strat_best_score(s, m, posl);
            pavel_score += score.second;
            s += score.first;
        }
        else{
            pair<int, int> score = strat_best_score(s, m, posl);
            vika_score += score.second;
            s += score.first;
        }
        //cout << s << " "  << pavel_score << " / " << vika_score << endl;
        turn++;
    }
    if (pavel_score > vika_score) {
        cout << "1" << endl; //"PAVEL WIN"
        //cout << pavel_score << " / " << vika_score << endl;
        return;
    }
    else {
        cout << "0" << endl; //"VIKA WIN"
        //cout << pavel_score << " / " << vika_score << endl;
        //return;
    }
    return;
}